src/reflect/scala/reflect/internal/util/StatisticsStatics.java

```diff
@@ -22,46 +21,15 @@
  * which helps performance (see docs to find out why).
  */
 public final class StatisticsStatics {
-  private static final AlmostFinalValue COLD_STATS = new AlmostFinalValue() {
-    @Override
-    protected boolean initialValue() {
-        return false;
-    }
-  };
-
-  private static final AlmostFinalValue HOT_STATS = new AlmostFinalValue() {
-    @Override
-    protected boolean initialValue() {
-        return false;
-    }
-  };
+  private static final AlmostFinalValue COLD_STATS = new AlmostFinalValue();
+  private static final AlmostFinalValue  HOT_STATS = new AlmostFinalValue();
 
   private static final MethodHandle COLD_STATS_GETTER = COLD_STATS.createGetter();
-  private static final MethodHandle HOT_STATS_GETTER = HOT_STATS.createGetter();
-  
-  public static boolean areSomeColdStatsEnabled() throws Throwable {
-    return (boolean) COLD_STATS_GETTER.invokeExact();
-  }
-
-  public static boolean areSomeHotStatsEnabled() throws Throwable {
-    return (boolean) HOT_STATS_GETTER.invokeExact();
-  }
-
-  public static void enableColdStats() throws Throwable {
-    if (!areSomeColdStatsEnabled())
-      COLD_STATS.setValue(true);
-  }
-
-  public static void disableColdStats() {
-    COLD_STATS.setValue(false);
-  }
+  private static final MethodHandle  HOT_STATS_GETTER = HOT_STATS.createGetter();
 
-  public static void enableHotStats() throws Throwable {
-    if (!areSomeHotStatsEnabled())
-      HOT_STATS.setValue(true);
-  }
+  public static boolean areSomeColdStatsEnabled() throws Throwable { return (boolean) COLD_STATS_GETTER.invokeExact(); }
+  public static boolean areSomeHotStatsEnabled()  throws Throwable { return (boolean)  HOT_STATS_GETTER.invokeExact(); }
 
-  public static void disableHotStats() {
-    HOT_STATS.setValue(false);
-  }
+  public static void enableColdStatsAndDeoptimize() { COLD_STATS.toggleOnAndDeoptimize(); }
```

Comment:
any reason these are not marked as `final`?

Reply:
Why should they be?  The class is final.

Reply:
Because a non-local change might break performance assumptions. In my experience, performance work is greatly helped by putting desired behavior as close as possible to the optimizations, and avoid depending on non-local settings—to avoid non-local changes have unforeseen impacts on optimized code(-paths). Also, different VMs might have different optimization strategies whether the class itself is final or not, and relying on implementation is difficult to judge the consequences of.

Reply:
Does making a static method final or even its class final, change the performance of an invocation of `StatisticsStatics.areSomeColdStatsEnabled`?  Otherwise that's just paranoid development and/or cargo culting.

If we're concerned with regressions I'd rather have something more robust than assumptions.